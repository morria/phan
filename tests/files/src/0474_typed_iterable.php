<?php

/**
 * @param iterable<string,int> $isi
 */
function expect_iterable_string_int($isi) {
    foreach ($isi as $k => $v) {
        echo strlen($v);  // should warn
        echo intdiv($k, 2);  // should also warn
    }
}

/**
 * @param Traversable<string, int> $tsi
 */
function expect_traversable_string_int(Traversable $tsi) {
    foreach ($tsi as $k => $v) {
        echo strlen($v);  // warn ($v is string)
        echo intdiv($k, 2);  // warn ($k is int)
    }
}
/**
 * @param Traversable<int,int> $tii
 * @param Traversable<string,int> $tsi
 * @param Traversable<string,string> $tss
 */
function expect_traversables(Traversable $tii, Traversable $tsi, Traversable $tss) {
    expect_traversable_string_int($tii);
    expect_traversable_string_int($tsi);
    expect_traversable_string_int($tss);
    // TODO: Make a subset of the below casts to iterable<TKey,TValue> emit appropriate warnings
    expect_iterable_string_int($tii);
    expect_iterable_string_int($tsi);
    expect_iterable_string_int($tss);
}

function test_iterable(string $s, int $i) {
    expect_iterable_string_int([$s => $i]);
    expect_iterable_string_int([$s => $s]);
    expect_iterable_string_int([$i => $s]);

    expect_traversable_string_int([$s => $i]);  // should warn
}
// TODO: Add tests of casting generators and subclasses of Traversable to iterable (and implement the functionality)

/**
 * @param Generator<string, int> $gsi
 */
function expect_generator_string_int(Generator $gsi) {
    foreach ($gsi as $k => $v) {
        echo strlen($v);  // warn ($v is string)
        echo intdiv($k, 2);  // warn ($k is int)
    }
}
