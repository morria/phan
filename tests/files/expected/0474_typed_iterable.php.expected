%s:8 PhanTypeMismatchArgumentInternal Argument 1 (string) is int but \strlen() takes string
%s:9 PhanTypeMismatchArgumentInternal Argument 1 (numerator) is string but \intdiv() takes int
%s:18 PhanTypeMismatchArgumentInternal Argument 1 (string) is int but \strlen() takes string
%s:19 PhanTypeMismatchArgumentInternal Argument 1 (numerator) is string but \intdiv() takes int
%s:28 PhanTypeMismatchArgument Argument 1 (tsi) is \Traversable|\Traversable<int,int>|iterable but \expect_traversable_string_int() takes \Traversable<string,int> defined at %s:16
%s:30 PhanTypeMismatchArgument Argument 1 (tsi) is \Traversable|\Traversable<string,string>|iterable but \expect_traversable_string_int() takes \Traversable<string,int> defined at %s:16
%s:39 PhanTypeMismatchArgument Argument 1 (isi) is array<string,string> but \expect_iterable_string_int() takes iterable<string,int> defined at %s:6
%s:40 PhanTypeMismatchArgument Argument 1 (isi) is array<int,string> but \expect_iterable_string_int() takes iterable<string,int> defined at %s:6
%s:42 PhanTypeMismatchArgument Argument 1 (tsi) is array<string,int> but \expect_traversable_string_int() takes \Traversable<string,int> defined at %s:16
%s:51 PhanTypeMismatchArgumentInternal Argument 1 (string) is int but \strlen() takes string
%s:52 PhanTypeMismatchArgumentInternal Argument 1 (numerator) is string but \intdiv() takes int
